vidarr-server/src/main/java/ca/on/oicr/gsi/vidarr/server/DatabaseBackedProcessor.java

```diff
@@ -429,12 +434,78 @@ private void addNewExternalKeyVersions(
     externalVersionInsert.execute();
   }
 
-  protected final <T> T delete(String workflowRunId, DeleteResultHandler<T> handler) {
+  protected final Result<Record4<String, String, String, String>>
+      getArgsForConsumableResourceRelease(DSLContext transaction, List<String> workflowRunHashIds) {
+    Long[] workflowRunIds =
+        transaction
+            .select(WORKFLOW_RUN.ID)
+            .from(WORKFLOW_RUN)
+            .where(WORKFLOW_RUN.HASH_ID.in(workflowRunHashIds))
+            .fetchArray(WORKFLOW_RUN.ID);
+    return getArgsForConsumableResourceRelease(transaction, workflowRunIds);
+  }
+
+  protected final Result<Record4<String, String, String, String>>
+      getArgsForConsumableResourceRelease(DSLContext transaction, Long[] workflowRunIds) {
+    return transaction
+        .select(
+            WORKFLOW_RUN.HASH_ID,
+            WORKFLOW_VERSION.NAME,
+            WORKFLOW_VERSION.VERSION,
+            ACTIVE_WORKFLOW_RUN.TARGET)
+        .from(WORKFLOW_VERSION)
+        .join(WORKFLOW_RUN)
+        .on(WORKFLOW_VERSION.ID.eq(WORKFLOW_RUN.WORKFLOW_VERSION_ID))
+        .join(ACTIVE_WORKFLOW_RUN)
+        .on(WORKFLOW_RUN.ID.eq(ACTIVE_WORKFLOW_RUN.ID))
+        .where(
+            WORKFLOW_RUN
+                .ID
+                .in(workflowRunIds)
+                .and(ACTIVE_WORKFLOW_RUN.ENGINE_PHASE.eq(Phase.WAITING_FOR_RESOURCES)))
+        .fetch();
+  }
+
+  protected final void releaseConsumableResources(
+      Map<String, Target> targets, Result<Record4<String, String, String, String>> args) {
+    args.forEach(
+        record -> {
+          final String workflowRunHashId = record.value1();
+          final String workflowName = record.value2();
+          final String workflowVersion = record.value3();
+          final String workflowRunTarget = record.value4();
+
+          targets
+              .get(workflowRunTarget)
+              .consumableResources()
+              .forEach(
+                  (b) ->
+                      b.second()
+                          .release(
+                              workflowName, workflowVersion, workflowRunHashId, Optional.empty()));
+        });
+  }
+
+  protected final <T> T delete(
+      String workflowRunId, Map<String, Target> targets, DeleteResultHandler<T> handler) {
     try (final Connection connection = dataSource.getConnection()) {
       return DSL.using(connection, SQLDialect.POSTGRES)
           .transactionResult(
               context -> {
                 final DSLContext transaction = DSL.using(context);
+                Result<Record4<String, String, String, String>> crArgs =
```

Comment:
crArgs is only used if we go into the `if(id_and_dead.component2())` block at 519. Could this call be moved there? It looks like id_and_dead.component1 is the workflow run id